fix: Incorrect ref count

arv opened the pull request:
If there is a diamond shape (or similar) we could end up writing a stale
ref count.

This happened because we read the ref count and then we do a bunch of
async operations. Then we write the new ref count based on the old, now
possible stale ref count value.

arv committed:
fix: Incorrect ref count

If there is a diamond shape (or similar) we could end up writing a stale
ref count.

This happened because we read the ref count and then we do a bunch of
async operations. Then we write the new ref count based on the old, now
possible stale ref count value.

arv commented on src/dag/write.ts:
```diff
@@ -85,64 +87,67 @@ export class Write extends Read {
-        await Promise.all(ps);
-      }
+  private async _changeRefCount(hash: Hash, delta: number): Promise<void> {
+    let lock = this._changeRefCountLocks.get(hash);
```
It might work to have a sync cache for the ref count instead. Trying that

aboodman replied:
Interesting and subtle bug. Yeah my mind went to same place as yours -- can we sort of parallelize getting all the ref counts that we will need in one pass, change them as needed, then flush them in a different pass? Would tha t simplify things?

arv replied:
That seems to work

arv committed:
Write ref counts to a map which we can read synchronously

arv committed:
Add a comment

aboodman commented on src/dag/write.ts:
```diff
@@ -85,75 +85,81 @@ export class Write extends Read {
-    } else {
-      await this.setRefCount(hash, newCount);
+    {
+      // Read again since the ref count might have changed.
```
This line is pretty scary to me. we read a value, make a decision base don that value (whether to descend) but then read it again because it might have changed?? 😬😬😬

Can I propose an alternate? What if we treat this like a breadth-first problem. Starting at the root:

1. collect all the ref changes to unique hashes in memory (so like if two different heads both reference same hash, then at this phase we would know that that hash is +2).
2. In parallel, execute all those changes.
3. For any hashes that went 0->1 or 1->0 do 1 again.

aboodman replied:
(alternately, I think the previous solution with a per-hash lock was also clearly correct, which I'm less sure about here).

arv committed:
Make the udpate of the ref count "atomic"

By only loading the ref count once, and after that only operate on the
cache we can ensure we are always working with the up to data ref count.
